server/src/test/java/io/druid/client/HttpServerInventoryViewTest.java

```diff
@@ -0,0 +1,334 @@
+            segmentAddLathces.get(segment.getIdentifier()).countDown();
+            return ServerView.CallbackAction.CONTINUE;
+          }
+
+          @Override
+          public ServerView.CallbackAction segmentRemoved(
+              DruidServerMetadata server, DataSegment segment
+          )
+          {
+            segmentDropLatches.get(segment.getIdentifier()).countDown();
+            return ServerView.CallbackAction.CONTINUE;
+          }
+
+          @Override
+          public ServerView.CallbackAction segmentViewInitialized()
+          {
+            initializeCallback1.countDown();
+            return ServerView.CallbackAction.CONTINUE;
+          }
+        }
+    );
+
+    final CountDownLatch serverRemovedCalled = new CountDownLatch(1);
+    httpServerInventoryView.registerServerRemovedCallback(
+        MoreExecutors.sameThreadExecutor(),
+        new ServerView.ServerRemovedCallback()
+        {
+          @Override
+          public ServerView.CallbackAction serverRemoved(DruidServer server)
+          {
+            if (server.getName().equals("host:8080")) {
+              serverRemovedCalled.countDown();
+              return ServerView.CallbackAction.CONTINUE;
+            } else {
+              throw new RE("Unknown server [%s]", server.getName());
+            }
+          }
+        }
+    );
+
+    httpServerInventoryView.start();
+
+    druidNodeDiscovery.listener.nodesAdded(ImmutableList.of(druidNode));
+
+    initializeCallback1.await();
+    segmentAddLathces.get(segment1.getIdentifier()).await();
+    segmentDropLatches.get(segment1.getIdentifier()).await();
+    segmentAddLathces.get(segment2.getIdentifier()).await();
+    segmentAddLathces.get(segment3.getIdentifier()).await();
+    segmentAddLathces.get(segment4.getIdentifier()).await();
+    segmentDropLatches.get(segment2.getIdentifier()).await();
+
+    DruidServer druidServer = httpServerInventoryView.getInventoryValue("host:8080");
+    Assert.assertEquals(ImmutableMap.of(segment3.getIdentifier(), segment3, segment4.getIdentifier(), segment4),
+                        druidServer.getSegments());
+
+    druidNodeDiscovery.listener.nodesRemoved(ImmutableList.of(druidNode));
+
+    serverRemovedCalled.await();
+    Assert.assertNull(httpServerInventoryView.getInventoryValue("host:8080"));
+
+    EasyMock.verify(druidNodeDiscoveryProvider);
+
+    httpServerInventoryView.stop();
+  }
+
+  private static class TestDruidNodeDiscovery implements DruidNodeDiscovery
+  {
+    Listener listener;
+
+    @Override
+    public Collection<DiscoveryDruidNode> getAllNodes()
+    {
+      throw new UnsupportedOperationException("Not Implemented.");
+    }
+
+    @Override
+    public void registerListener(Listener listener)
+    {
+      listener.nodesAdded(ImmutableList.of());
+      this.listener = listener;
+    }
+  }
+
+  private static class TestHttpClient implements HttpClient
+  {
+    BlockingQueue<ListenableFuture> results;
+    int requestNum = 0;
```

Comment:
Does this should be AtomicInteger?

Reply:
yeah, that would make it more robust.. changed.